### PR TITLE
Parse friending

### PR DESCRIPTION
**Pull request by s-fukuzawaa**
**### Friending**

- Implemented friending backend.
- User can friend request a user
- User can unfriend a user
- User can accept or reject a user
- User can see their follow requests in their activities page

Kapture 2022-07-15 at 17 26 13

### CODE REVIEW

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m [hunk 1]
```diff
@@ -6,41 +6,77 @@
 //
 
 #import "ActivitiesViewController.h"
-
-@interface ActivitiesViewController ()<UITableViewDataSource, UITableViewDelegate>
+#import "FriendProfileViewController.h"
+#import "ActivityCell.h"
+#import "Parse/Parse.h"
+#import "AlbumConstants.h"
+@interface ActivitiesViewController ()<UITableViewDataSource, UITableViewDelegate, ActivityCellDelegate>
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:13]: lets add a empty line after the import

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m [hunk 2]
```diff
@@ -6,41 +6,77 @@
...
+	self.requests = [[NSMutableArray alloc] init];
+	PFQuery *query = [PFQuery queryWithClassName:classNameFriendship];
+	PFUser *currentUser = [PFUser currentUser];
+	[query whereKey:@"recipientId" equalTo:currentUser.objectId];
+	[query whereKey:@"hasFriended" equalTo:@(PENDING)];
+	[query orderByDescending:@"updatedAt"];
+	[query findObjectsInBackgroundWithBlock:^(NSArray *friendships, NSError *error) {
+	         if (friendships != nil) {
+			 for(PFObject *friendship in friendships) {
+				 PFQuery *query = [PFUser query];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:52]: you create d a new query, but we are never calling the query. 
How come query contains the object `user` with [query findObjects][0]?

File: Album/ViewControllers/ProfileControllers/AddFriendViewController.m [hunk 3]
```diff
@@ -20,47 +20,48 @@ @interface AddFriendViewController () <UITableViewDataSource, UITableViewDelegat
...
+	PFQuery *query = [PFUser query];
+	PFUser *currentUser = [PFUser currentUser];
+	[query whereKey:@"username" notEqualTo:currentUser.username];
+	[query whereKey:@"username" equalTo:self.searchFriendField.text];
+	[query findObjectsInBackgroundWithBlock:^(NSArray *users, NSError *error) {
+	         if (users != nil) {
+			 self.friendsArr = (NSMutableArray*)users;
+			 // Set up table view
+			 self.tableView.dataSource = self;
+			 self.tableView.delegate = self;
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/AddFriendViewController.m:36]: This does not seems the right place to set the `dataSource` and the `delegate`

File: Album/ViewControllers/ProfileControllers/FriendProfileViewController.m [hunk 4]
```diff
@@ -6,76 +6,263 @@
...
+- (void) updateButton {
+	UIColor *friendshipButtonBackgroundColor;
+	NSString *friendshipButtonText;
+	UIColor *friendshipButtonTitleColor;
+	if(self.request != NULL && [self.requestStatus intValue]== PENDING) {
+		friendshipButtonBackgroundColor = [UIColor whiteColor];
+		friendshipButtonText = @"Received Request";
+		friendshipButtonTitleColor = [UIColor colorWithRed: 0.39 green: 0.28 blue: 0.22 alpha: 1.00];
+	}else if([self.friendStatus intValue] == FRIENDED) {
+		friendshipButtonBackgroundColor = [UIColor colorWithRed: 0.39 green: 0.28 blue: 0.22 alpha: 1.00];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/FriendProfileViewController.m:78]: that color seems like it belongs to a contant.

File: Album/ViewControllers/ProfileControllers/FriendProfileViewController.m [hunk 5]
```diff
@@ -6,76 +6,263 @@
...
+			 }else{
+				 // Otherwise, store
+				 self.friendship = friendships[0];
+				 self.friendStatus = self.friendship.hasFriended;
+				 // Update button UI
+				 [self updateButton];
+			 }
+
+		 } else {
+			 NSLog(@"%@", error.localizedDescription);
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/FriendProfileViewController.m:118]: should we alert an error?

File: Album/ViewControllers/ProfileControllers/FriendProfileViewController.m [hunk 6]
```diff
@@ -6,76 +6,263 @@
...
+	                         style:UIAlertActionStyleDefault
+	                         handler: nil];
+	[alert addAction:cancel];
+	[self presentViewController:alert animated:YES completion:nil];
 }
 
-- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
-    return NULL;
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
+	return 0;
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/FriendProfileViewController.m:262]: so much better

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m [hunk 7]
```diff
@@ -22,71 +22,71 @@ @interface ProfileViewController () <UICollectionViewDataSource, UICollectionVie
...
-- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
-    return 0;
+- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
+	return 0;
 }
 
-- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
-    return CGSizeMake(3, 3);
+- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
+	return CGSizeMake(3, 3);
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ProfileViewController.m:47]: 3 by 3 pixels?

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m [hunk 8]
```diff
@@ -22,71 +22,71 @@ @interface ProfileViewController () <UICollectionViewDataSource, UICollectionVie
...
+		UINavigationController *navigationController = [segue destinationViewController];
+		ActivitiesViewController *activitiesController = (ActivitiesViewController*)navigationController.topViewController;
+		//        activitiesController.delegate = self; TODO: Add delegate later
+
+	}else if([segue.identifier isEqualToString:@"addFriendSegue"]) {
+		UINavigationController *navigationController = [segue destinationViewController];
+		AddFriendViewController *addFriendController = (AddFriendViewController*)navigationController.topViewController;
+		//        addFriendController.delegate = self; TODO: Add delegate later
+	}else if([segue.identifier isEqualToString:@"settingsSegue"]) {
+		UINavigationController *navigationController = [segue destinationViewController];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ProfileViewController.m:86]: seems like:
```
UINavigationController *navigationController = [segue destinationViewController];
```
is repeated on every if-statement

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m [hunk 9]
```diff
@@ -21,32 +21,22 @@ @interface SettingsViewController ()
 @implementation SettingsViewController
 
 - (void)viewDidLoad {
-    [super viewDidLoad];
-    // Do any additional setup after loading the view.
+	[super viewDidLoad];
+	// Do any additional setup after loading the view.
 }
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/SettingsViewController.m:26]: we could remove this

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m [hunk 10]
```diff
@@ -21,32 +21,22 @@ @interface SettingsViewController ()
...
- // In a storyboard-based application, you will often want to do a little preparation before navigation
- - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
- // Get the new view controller using [segue destinationViewController].
- // Pass the selected object to the new view controller.
- }
- */
+	UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
+	LoginViewController *loginViewController = [storyboard instantiateViewControllerWithIdentifier:@"LoginViewController"];
+	myDelegate.window.rootViewController = loginViewController;
+	[PFUser logOutInBackgroundWithBlock:^(NSError * _Nullable error) {
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/SettingsViewController.m:38]: if the block is empty, maybe we could pass `nil`.

**Review by albertomota-codes — APPROVED** [review]

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,2 @@
-
+xcuserdata/eyeray.xcuserdatad/
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply .gitignore]: I don't think it did...?

File: Album/AppDelegate.m [hunk 12]
```diff
@@ -29,8 +29,8 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     NSString *clientKey = [dict objectForKey: @"client_key"];
     ParseClientConfiguration *config = [ParseClientConfiguration  configurationWithBlock:^(id<ParseMutableClientConfiguration> configuration) {
 
-            configuration.applicationId = appId; // <- UPDATE
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/AppDelegate.m:32]: That update was comment from CodePath where it meant like "update default app id with your own." We're good here

File: Album/Models/Friendship.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+#import <Parse/Parse.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface Friendship : PFObject<PFSubclassing>
+@property (nonatomic, strong) NSString *objectId;
+@property (nonatomic, strong) NSString *requesterId;
+@property (nonatomic, strong) NSString *recipientId;
+@property (nonatomic, strong) NSNumber *hasFriended;
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Models/Friendship.h:16]: Yes it was. But considering how there should be pending status, I figured BOOL was not binary so I made it into NSNumber.

File: Album/Models/Friendship.m [hunk 14]
```diff
@@ -0,0 +1,19 @@
+//
+//  Friendship.m
+//  Album
+//
+//  Created by Airei Fukuzawa on 7/14/22.
+//
+
+#import "Friendship.h"
+#import "AlbumConstants.h"
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Models/Friendship.m:9]: Will do!

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m (hunk 7 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ProfileViewController.m:47]: This is a temporary place holder. Will change later when finish implementing profile views.

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m (hunk 8 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ProfileViewController.m:86]: Oops. Refactored. Thank you!

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m (hunk 9 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/SettingsViewController.m]: Done!

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m (hunk 10 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/SettingsViewController.m:37]: Agree. Done!

File: Album/Views/ActivityCell.m [hunk 15]
```diff
@@ -6,9 +6,36 @@
 //
 
 #import "ActivityCell.h"
-
 @implementation ActivityCell
+- (void)setUser:(PFUser *)user {
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Views/ActivityCell.m:11]: Done!

File: Album.xcworkspace/xcuserdata/eyeray.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist [hunk 16]
```diff
@@ -7,336 +7,32 @@
       <BreakpointProxy
          BreakpointExtensionID = "Xcode.Breakpoint.FileBreakpoint">
          <BreakpointContent
-            uuid = "6403575F-542E-4595-A6EB-554F86371446"
+            uuid = "1EEF7E71-730C-4E3D-8044-E78CBA1CBDFD"
             shouldBeEnabled = "No"
             ignoreCount = "0"
             continueAfterRunningActions = "No"
-            filePath = "Album/ViewControllers/ProfileControllers/ProfileViewController.m"
```
Comment (albertomota-codes) [inline Album.xcworkspace/xcuserdata/eyeray.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:14]: At some point we need to remove this .xbkptlist from the repo. 
We do not want reviewers to see this.

File: Album/AppDelegate.m [hunk 17]
```diff
@@ -8,6 +8,7 @@
 #import "AppDelegate.h"
 #import "Parse/Parse.h"
 @import GoogleMaps;
+
```
Comment (albertomota-codes) [inline Album/AppDelegate.m:11]: 💙

File: Album/Models/Friendship.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (albertomota-codes) [reply Album/Models/Friendship.h:16]: I would recommend to name it like `friendStatus`. 

the `has____`  or `is_____` format most of the times indicates it is a BOOL.

[the remark this answers is not in the document]
Comment (albertomota-codes) [reply Album/Models/Friendship.h:16]: even better if we could have an `enum` instead of a NSNumber.

**Pull request merged**